third_party/move/move-vm/runtime/src/runtime.rs

```diff
@@ -394,6 +394,41 @@ impl VMRuntime {
         extensions: &mut NativeContextExtensions,
         bypass_declared_entry_check: bool,
     ) -> VMResult<SerializedReturnValues> {
+        // load the function
+        let (module, func, instantiation) =
+            self.loader
+                .load_function(module, function_name, &ty_args, data_store)?;
+
+        self.execute_function_instantiation(
+            module,
+            func,
+            instantiation,
+            serialized_args,
+            data_store,
+            gas_meter,
+            extensions,
+            bypass_declared_entry_check,
+        )
+    }
+
+    pub(crate) fn execute_function_instantiation(
+        &self,
+        module: Arc<Module>,
+        function: Arc<Function>,
```

Comment:
Again, I understand that you're trying to avoid doing duplicated work, but Module & Function are internal types that we should try not to expose. Let me see if I can propose an alternative set of APIs that would allow us to hide these types.

Reply:
How about
```
pub fn infer_type_parameters_by_return_type(&self, expected) -> BTreeMap<usize, TypeTag>;
```
Internally, you can infer the types as usual and convert them to TypeTag using `fn type_to_type_tag`. The client will simply call the existing `fn execute_function_bypass_visibility`. I know it's doing a bit of duplicated work, but this way we can avoid leaking the internal types of the loader.

Reply:
I implemented the wrapper struct as proposed offline, PTAL

Reply:
LGTM. Would be great if you could add a few comments documenting the constraint that `LoadedFunction` must be treated as an identifier rather than an owned object.